velox/type/TimestampConversion.cpp

```diff
@@ -812,6 +814,83 @@ Status parserError(const char* str, size_t len) {
       std::string(str, len));
 }
 
+inline bool startsWith(std::string_view str, const char* prefix) {
+  return str.rfind(prefix, 0) == 0;
+}
+
+bool parseSparkTzWithPrefix(
+    std::string& timeZoneName,
+    Timestamp& ts,
+    size_t start) {
+  VELOX_DCHECK_GT(timeZoneName.length(), start + 2);
+  auto sign = timeZoneName[start];
+  auto offset = timeZoneName.substr(start + 1);
+  std::string hm, h, m, s;
+  RE2 hhPattern(R"((\d?\d))");
+  RE2 hhmmPattern(R"((\d\d):?(\d\d))");
+  RE2 hhmmssPattern(R"((\d\d:\d\d):(\d\d))");
+  RE2 hhmmssCompactPattern(R"((\d\d)(\d\d)(\d\d))");
+  if (RE2::FullMatch(offset, hhPattern, &hm)) {
+    hm += ":00";
+  } else if (RE2::FullMatch(offset, hhmmPattern, &h, &m)) {
+    hm = h + ":" + m;
+  } else if (RE2::FullMatch(offset, hhmmssPattern, &hm, &s)) {
+    // hm and second already captured
+  } else if (RE2::FullMatch(offset, hhmmssCompactPattern, &h, &m, &s)) {
+    hm = h + ":" + m;
+  } else {
+    return false;
+  }
+  int32_t sec = 0;
+  size_t pos = 0;
+  if (!s.empty() && !parseDoubleDigit(s.c_str(), s.length(), pos, sec)) {
+    return false;
+  }
+  if (sec < 0 || sec > 60) {
+    return false;
+  }
+  if (sec != 0) {
+    // Seconds timezone offset is not support, need to be applied to timestamp.
+    if (sign == '+') {
+      if (ts.getSeconds() - sec < Timestamp::kMinSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() - sec, ts.getNanos());
+    } else if (sign == '-') {
+      if (ts.getSeconds() + sec > Timestamp::kMaxSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() + sec, ts.getNanos());
+    } else {
+      return false;
+    }
+  }
+  timeZoneName = sign + hm;
+  return true;
+}
+
+bool normalizeSparkTimezone(std::string& timeZoneName, Timestamp& ts) {
+  // Spark support an id with one of the prefixes UTC+, UTC-, GMT+, GMT-, UT+ or
+  // UT-, and a suffix in the formats: h[h], hh[:]mm, hh:mm:ss, hhmmss
+  if (timeZoneName.length() > 5 &&
+      (startsWith(timeZoneName, "UTC") || startsWith(timeZoneName, "GMT"))) {
+    if (!parseSparkTzWithPrefix(timeZoneName, ts, 3)) {
+      return false;
+    }
+  }
+  if (timeZoneName.length() > 4 && startsWith(timeZoneName, "UT")) {
+    if (!parseSparkTzWithPrefix(timeZoneName, ts, 2)) {
+      return false;
+    }
+  }
+  RE2 singleHourTz(R"((\+|\-)(\d):)");
```

Comment:
we have similar logic to parse timezone offsets (full or partially defined). Could we reuse it here instead of relying on re2?

Reply:
In this implementation, I followed the handling from Spark's source code, where Spark uses regex matching to pad the hours and minutes to two digits. Therefore, I used re2, and I will try to reuse the existing code. I will also attempt to modify other parts that use re2.

Reply:
@pedroerp I checked the `TimeZoneMap::normalizeTimeZoneOffset` method, and its normalization rules are inconsistent with Spark's rules in this context. Spark corrects timezone offsets like `+2:3`, `+02:3`, or `+2:03` to `+02:03`. I will add the normalization rule in Spark to not use re2.